packages/autocomplete-core/src/onKeyDown.ts

```diff
@@ -22,39 +23,72 @@ export function onKeyDown<TItem extends BaseItem>({
   ...setters
 }: OnKeyDownOptions<TItem>): void {
   if (event.key === 'ArrowUp' || event.key === 'ArrowDown') {
-    // Default browser behavior changes the caret placement on ArrowUp and
-    // Arrow down.
-    event.preventDefault();
+    // eslint-disable-next-line no-inner-declarations
+    function triggerScrollIntoView() {
+      const nodeItem = props.environment.document.getElementById(
+        `${props.id}-item-${store.getState().activeItemId}`
+      );
 
-    store.dispatch(event.key, null);
+      if (nodeItem) {
+        if ((nodeItem as any).scrollIntoViewIfNeeded) {
+          (nodeItem as any).scrollIntoViewIfNeeded(false);
+        } else {
+          nodeItem.scrollIntoView(false);
+        }
+      }
+    }
 
-    const nodeItem = props.environment.document.getElementById(
-      `${props.id}-item-${store.getState().activeItemId}`
-    );
+    // eslint-disable-next-line no-inner-declarations
+    function triggerOnActive() {
+      const highlightedItem = getActiveItem(store.getState());
 
-    if (nodeItem) {
-      if ((nodeItem as any).scrollIntoViewIfNeeded) {
-        (nodeItem as any).scrollIntoViewIfNeeded(false);
-      } else {
-        nodeItem.scrollIntoView(false);
+      if (store.getState().activeItemId !== null && highlightedItem) {
+        const { item, itemInputValue, itemUrl, source } = highlightedItem;
+
+        source.onActive({
+          event,
+          item,
+          itemInputValue,
+          itemUrl,
+          refresh,
+          source,
+          state: store.getState(),
+          ...setters,
+        });
       }
     }
 
-    const highlightedItem = getActiveItem(store.getState());
-
-    if (store.getState().activeItemId !== null && highlightedItem) {
-      const { item, itemInputValue, itemUrl, source } = highlightedItem;
+    // Default browser behavior changes the caret placement on ArrowUp and
+    // Arrow down.
+    event.preventDefault();
 
-      source.onActive({
+    // When re-opening the panel, we need to split the logic to keep the actions
+    // synchronized as `onInput` returns a promise.
+    if (
+      store.getState().isOpen === false &&
+      (props.openOnFocus || Boolean(store.getState().query))
+    ) {
+      onInput({
         event,
-        item,
-        itemInputValue,
-        itemUrl,
+        props,
+        query: store.getState().query,
         refresh,
-        source,
-        state: store.getState(),
+        store,
         ...setters,
+      }).then(() => {
+        store.dispatch(event.key as ActionType, {
+          nextActiveItemId: props.defaultActiveItemId,
+        });
+
+        // We need to wait for the panel to open.
+        setTimeout(triggerScrollIntoView, 0);
+        triggerOnActive();
       });
+    } else {
+      store.dispatch(event.key, {});
+
+      triggerScrollIntoView();
+      triggerOnActive();
```

Comment:
To keep the same order as above now:

```suggestion
      triggerOnActive();
      triggerScrollIntoView();
```

Reply:
Fixed in [`120f6dc`](https://github.com/algolia/autocomplete/pull/599/commits/120f6dc4d8f2351f9f28fb4228eb585d9d488898#diff-a33bd5a108ed1cc854e9a594b5ec2bef4313ea5d549ddd73473f841e35bed681R83-R93)